landoapi/hg.py

```diff
@@ -150,6 +202,82 @@ def clean_repo(self, strip_non_public_commits=True):
             except hglib.error.CommandError:
                 pass
 
+    def apply_patch(self, patch_io_buf):
+        patch_helper = PatchHelper(patch_io_buf)
+        if not patch_helper.diff_start_line:
+            raise NoDiffStartLine()
+
+        # Import the diff to apply the changes then commit separately to
+        # ensure correct parsing of the commit message.
+        f_msg = tempfile.NamedTemporaryFile()
+        f_diff = tempfile.NamedTemporaryFile()
+        with f_msg, f_diff:
+            patch_helper.write_commit_description(f_msg)
+            f_msg.flush()
+            patch_helper.write_diff(f_diff)
+            f_diff.flush()
+
+            # TODO: Using `hg import` here is less than ideal because
+            # it does not use a 3-way merge. It would be better
+            # to use `hg import --exact` then `hg rebase`, however we
+            # aren't guaranteed to have the patche's parent changeset
+            # in the local repo.
+            import_cmd = ["import", "--no-commit"]
+            # Apply the patch, with file rename detection (similarity).
+            # Using 95 as the similarity to match automv's default.
+            import_cmd += ["-s", "95"]
+
+            try:
+                self.run_hg(import_cmd + [f_diff.name])
+            except hglib.error.CommandError as exc:
+                if isinstance(HgException.from_hglib_error(exc), PatchConflict):
+                    # Try again using 'patch' instead of hg's internal patch utility.
+                    # But first reset to a clean working directory as hg's attempt
+                    # might have partially applied the patch.
+                    logger.info("import failed, retrying with 'patch'", exc_info=exc)
+                    import_cmd += ["--config", "ui.patch=patch"]
+                    self.clean_repo(strip_non_public_commits=False)
+
+                    try:
+                        # When using an external patch util mercurial won't
+                        # automatically handle add/remove/renames.
+                        cmd = import_cmd + [f_diff.name, "addremove", "-s", "95"]
+                        self.run_hg(cmd)
+                    except hglib.error.CommandError:
+                        # Use the original exception from import with the built-in
+                        # patcher since both attempts failed.
+                        raise HgException.from_hglib_error(exc) from exc
+
+            # Commit using the extracted date, user, and commit desc.
+            # --landing_system is provided by the set_landing_system hgext.
+            self.run_hg(
+                ["commit"]
+                + ["--date", patch_helper.header("Date")]
+                + ["--user", patch_helper.header("User")]
+                + ["--landing_system", "lando"]
+                + ["--logfile", f_msg.name]
+            )
+
+            # TODO: do we need to ever add "--encoding utf-8" to th commit command?
```

Comment:
Shouldn't be needed - we're setting the default encoding when starting up the hg command server.